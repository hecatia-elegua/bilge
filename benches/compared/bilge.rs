use bilge::{Bitsized, Number, FromBits, bitsize, u4, u12, u20, u5, u2, DebugBits};

#[inline(never)]
pub fn bilge(input: (u32, u32, u64, u16)) {
    let mut lpi = GicRedistributorLpi {
        control: RedistributorControl::from(input.0),
        implementer_identification: RedistributorImplementerIdentification::from(input.1),
        redistributor_type: RedistributorType::from(input.2),
    };
    assert_eq!(lpi.control.value, input.0);
    assert_eq!(lpi.implementer_identification.value, input.1);
    assert_eq!(lpi.redistributor_type.value, input.2);
    
    assert!(lpi.control.clear_enable_supported());
    assert_eq!(lpi.implementer_identification.implementer_jep106(), u12::new(2054));
    lpi.implementer_identification.set_implementer_jep106(u12::new(input.3));
    assert_eq!(lpi.implementer_identification.implementer_jep106(), u12::new(input.3));
    assert_eq!(lpi.redistributor_type.processor_number(), 63872);
}

#[derive(Debug)]
pub struct GicRedistributorLpi {
    control: RedistributorControl,
    implementer_identification: RedistributorImplementerIdentification,
    redistributor_type: RedistributorType,
}

#[bitsize(32)]
#[derive(DebugBits, FromBits)]
struct RedistributorControl {
    //ro
    upstream_write_pending: bool,
    //zero
    reserved: u4,
    //or reserved
        disable_processor_selection_for_group_1_secure_interrupts: bool,
        disable_processor_selection_for_group_1_non_secure_interrupts: bool,
        disable_processor_selection_for_group_0_interrupts: bool,
    //zero
    reserved: u20,
    register_write_pending: bool,
    lpi_invalidate_registers_supported: bool,
    clear_enable_supported: bool,
    enable_lpis: bool,
}

#[bitsize(32)]
#[derive(DebugBits, FromBits)]
struct RedistributorImplementerIdentification {
    //ro
    product_id: u8,
    //zero
    reserved: u4,
    //ro
    variant: u4,
    //ro
    revision: u4,
    implementer_jep106: u12,
}

#[bitsize(64)]
#[derive(DebugBits, FromBits)]
struct RedistributorType {
    affinity_value: u32,
    ppi_num: u5,
    virtual_sgi_supported: bool,
    common_lpi_affinity: u2, //TODO: enums everywhere
    processor_number: u16,
    resident_vpe_id: bool,
    mpam_supported: bool,
    control_cpgs_supported: bool,
    is_last: bool,
    direct_lpi_supported: bool,
    dirty: bool,
    virtual_lpi_supported: bool,
    physical_lpi_supported: bool,
}
