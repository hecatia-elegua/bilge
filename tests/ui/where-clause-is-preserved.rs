#![feature(const_trait_impl)]

use bilge::prelude::*;
use core::marker::PhantomData;

#[derive(Default)]
struct Generic<T>(u2, PhantomData<T>);

impl<T> Bitsized for Generic<T> {
    type ArbitraryInt = u2;
    const BITS: usize = 2;
    const MAX: Self::ArbitraryInt = <u2 as Bitsized>::MAX;
}

impl<T> const From<u2> for Generic<T> {
    fn from(val: u2) -> Self {
        Self(val, PhantomData)
    }
}

impl<T> const From<Generic<T>> for u2 {
    fn from(val: Generic<T>) -> u2 {
        val.0
    }
}

#[bitsize(5)]
struct BoundedGeneric<T>(Generic<T>, u3)
where
    T: Iterator<Item = ()>;

fn main() {
    // kinda ick, but the only way I could find to check that the clause was on the struct directly
    BoundedGeneric::<()> {
        value: <_>::default(),
        _phantom: PhantomData,
    };
    // check that where clause is present on inherent impl
    BoundedGeneric::<()>::new(<_>::default(), u3::new(0));
}
