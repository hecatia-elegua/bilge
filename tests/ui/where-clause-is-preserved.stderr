error[E0277]: `()` is not an iterator
  --> tests/ui/where-clause-is-preserved.rs:34:22
   |
34 |     BoundedGeneric::<()> {
   |                      ^^ `()` is not an iterator
   |
   = help: the trait `Iterator` is not implemented for `()`
note: required by a bound in `BoundedGeneric`
  --> tests/ui/where-clause-is-preserved.rs:30:8
   |
28 | struct BoundedGeneric<T>(Generic<T>, u3)
   |        -------------- required by a bound in this
29 | where
30 |     T: Iterator<Item = ()>;
   |        ^^^^^^^^^^^^^^^^^^^ required by this bound in `BoundedGeneric`

error[E0599]: the function or associated item `new` exists for struct `BoundedGeneric<()>`, but its trait bounds were not satisfied
  --> tests/ui/where-clause-is-preserved.rs:39:27
   |
27 | #[bitsize(5)]
   | ------------- function or associated item `new` not found for this struct
...
39 |     BoundedGeneric::<()>::new(<_>::default(), u3::new(0));
   |                           ^^^ function or associated item cannot be called on `BoundedGeneric<()>` due to unsatisfied trait bounds
   |
   = note: the following trait bounds were not satisfied:
           `<() as Iterator>::Item = ()`
           `(): Iterator`

error[E0277]: `()` is not an iterator
  --> tests/ui/where-clause-is-preserved.rs:39:5
   |
39 |     BoundedGeneric::<()>::new(<_>::default(), u3::new(0));
   |     ^^^^^^^^^^^^^^^^^^^^ `()` is not an iterator
   |
   = help: the trait `Iterator` is not implemented for `()`
note: required by a bound in `BoundedGeneric`
  --> tests/ui/where-clause-is-preserved.rs:30:8
   |
28 | struct BoundedGeneric<T>(Generic<T>, u3)
   |        -------------- required by a bound in this
29 | where
30 |     T: Iterator<Item = ()>;
   |        ^^^^^^^^^^^^^^^^^^^ required by this bound in `BoundedGeneric`
